pkg/crypto11/session.go

```diff
@@ -0,0 +1,133 @@
+package crypto11
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/miekg/pkcs11"
+	"github.com/pkg/errors"
+)
+
+// errTokenNotFound represents the failure to find the requested PKCS#11 token
+var errTokenNotFound = errors.New("could not find PKCS#11 token")
+
+type Session struct {
+	Ctx    *pkcs11.Ctx
+	Handle pkcs11.SessionHandle
+}
+
+type Config struct {
+	Path       string
+	TokenLabel string
+	Slot       *int
+	Pin        string
+}
+
+func NewSession(cfg *Config) (*Session, error) {
+	p := pkcs11.New(cfg.Path)
+	if p == nil {
+		return nil, fmt.Errorf("cannot create HSM access for PKCS#11")
+	}
+	err := p.Initialize()
+	if err != nil {
+		return nil, err
+	}
+
+	slots, err := lookupSlots(p)
+	if err != nil {
+		p.Destroy()
+		return nil, fmt.Errorf("lookup HSM slots: %w", err)
+	}
+
+	slot, err := findSlot(cfg, p, slots)
+	if err != nil {
+		p.Destroy()
+		return nil, fmt.Errorf("lookup HSM slots: %w", err)
+	}
```

Comment:
looking at the implementation of find slots
```
func findSlot(cfg *Config, p *pkcs11.Ctx, slots []uint) (uint, error) {
	if cfg.Slot == nil && cfg.TokenLabel == "" {
		slots, err := lookupSlots(p)
		if err != nil {
			return 0, fmt.Errorf("lookup HSM slots: %w", err)
		}
		return slots[0], nil
	}
	for _, slot := range slots {
...
```

the `lookupSlots(p)` call seems to be redundant. Besides I'm confused why within the `findSlot(...)`, p.Destroy() would not called in the error case?

Reply:
yes, I removed this, and return an error if there are no slots given.

Destroy is called by the caller at the top level function, when it is clear that ii is not required anymore.